tests/test_sig_stfl.c

```diff
@@ -0,0 +1,271 @@
+	rv |= memcmp(public_key - sizeof(magic_t), magic.val, sizeof(magic_t));
+	//rv |= memcmp(secret_key- sizeof(magic_t), magic.val, sizeof(magic_t));
+	rv |= memcmp(message - sizeof(magic_t), magic.val, sizeof(magic_t));
+	rv |= memcmp(signature - sizeof(magic_t), magic.val, sizeof(magic_t));
+	if (rv) {
+		fprintf(stderr, "ERROR: Magic numbers do not match\n");
+		goto err;
+	}
+#endif
+
+	printf("verification passes as expected\n");
+	ret = OQS_SUCCESS;
+	goto cleanup;
+
+err:
+	ret = OQS_ERROR;
+
+cleanup:
+	if (sig != NULL) {
+		OQS_SECRET_KEY_free(secret_key);
+	}
+	OQS_MEM_insecure_free(public_key - sizeof(magic_t));
+	OQS_MEM_insecure_free(message - sizeof(magic_t));
+	OQS_MEM_insecure_free(signature - sizeof(magic_t));
+
+	OQS_SIG_STFL_free(sig);
+
+	return ret;
+}
+
+#ifdef OQS_ENABLE_TEST_CONSTANT_TIME
+static void TEST_SIG_randombytes(uint8_t *random_array, size_t bytes_to_read) {
+	// We can't make direct calls to the system randombytes on some platforms,
+	// so we have to swap out the OQS_randombytes provider.
+	OQS_randombytes_switch_algorithm("system");
+	OQS_randombytes(random_array, bytes_to_read);
+	OQS_randombytes_custom_algorithm(&TEST_SIG_randombytes);
+
+	// OQS_TEST_CT_CLASSIFY tells Valgrind's memcheck tool to issue a warning if
+	// the program branches on any byte that depends on random_array. This helps us
+	// identify timing side-channels, as these bytes often contain secret data.
+	OQS_TEST_CT_CLASSIFY(random_array, bytes_to_read);
+}
+#endif
+
+#if OQS_USE_PTHREADS_IN_TESTS
+struct thread_data {
+	char *alg_name;
+	OQS_STATUS rc;
+};
+
+void *test_wrapper(void *arg) {
+	struct thread_data *td = arg;
+	td->rc = sig_test_correctness(td->alg_name);
+	return NULL;
+}
+#endif
+
+int main(int argc, char **argv) {
+
+	if (argc != 2) {
+		fprintf(stderr, "Usage: test_sig algname\n");
+		fprintf(stderr, "  algname: ");
+		for (size_t i = 0; i < OQS_SIG_STFL_algs_length; i++) {
+			if (i > 0) {
+				fprintf(stderr, ", ");
+			}
+			fprintf(stderr, "%s", OQS_SIG_STFL_alg_identifier(i));
+		}
+		fprintf(stderr, "\n");
+		return EXIT_FAILURE;
+	}
+
+	print_system_info();
+
+	char *alg_name = argv[1];
+	if (!OQS_SIG_STFL_alg_is_enabled(alg_name)) {
+		printf("Signature algorithm %s not enabled!\n", alg_name);
+		return EXIT_FAILURE;
+	}
+
+#ifdef OQS_ENABLE_TEST_CONSTANT_TIME
+	OQS_randombytes_custom_algorithm(&TEST_SIG_randombytes);
+#else
+	OQS_randombytes_switch_algorithm("system");
+#endif
+
+	OQS_STATUS rc;
+#if OQS_USE_PTHREADS_IN_TESTS
+#define MAX_LEN_SIG_NAME_ 64
+	// don't run Rainbow III and V in threads because of large stack usage
+	char no_thread_sig_patterns[][MAX_LEN_SIG_NAME_]  = {"Rainbow-III", "Rainbow-V"};
+	int test_in_thread = 1;
+	for (size_t i = 0 ; i < sizeof(no_thread_sig_patterns) / MAX_LEN_SIG_NAME_; ++i) {
+		if (strstr(alg_name, no_thread_sig_patterns[i]) != NULL) {
+			test_in_thread = 0;
+			break;
+		}
+	}
+	if (test_in_thread) {
```

Comment:
No need for any of these checks because Rainbow is not a stateful signature scheme.